236 batch delete

neiljohari opened the pull request:
In this PR we address issue https://github.com/eecs-autograder/ag-website-vue/issues/236 which describes the ability for instructors to delete available instructor files in batch instead of one-by-one. 

We preserve the single file deletion ability and additionally allow instructors to delete single files while still in batch selection mode.

This is accomplished by lifting the delete functionality and confirmation modal out of SingleInstructorFile to the parent component of InstructorFiles. The parent component manages generic deletion logic that sends delete requests asynchronously for an array of files.  When the user requests single file deletion or batch file deletion, the generic deletion mechanism is set up and then deferred to. This allows us to reduce code duplication while adding this functionality.

Note on tests: the current `develop` branch has a failing `InstructorFiles` test that this PR does not resolve

Fixes #236

neiljohari committed:
Add checkboxes and event for file selection

Begin work for batched operations by allowing users to select
SingleInstructorFiles. The child component emits a selected event now
to notify the parent when a user has marked the child for a batch
operation.

The list of selected files is stored in the parent component,
InstructorFiles, who listens for the select event.

Co-authored-by: Cayetano Wagner <cowagner@umich.edu>

neiljohari committed:
Add delete and select all button

neiljohari committed:
Implement batch delete instructor files

Lifted SingleInstructorFile delete functionality into InstructorFiles
component. Allows users to delete single files while in the middle of a
batch delete operations. Deletion is done in parallel with async
requests. Select all functionality remains to be implemented.

Co-authored-by: Cayetano Wagner <wagnerc1822@gmail.com>

caywagner committed:
Add singe and batch delete tests, remove old deletion test

caywagner committed:
Remove select all SingleInstructorFile checkbox

caywagner committed:
Prevent SingleInstructorFile checkbox to click through and display file

neiljohari committed:
Format code to meet quality requirements

james-perretta commented:
Thanks for this! Here's my first round of feedback on functionality/UI:

**High Priority**
- The "delete" button should be hidden when the uploaded files sidebar is collapsed (click on the hamburger icon to see this)
- When I run the tests locally on develop, there aren't any failures. It seems like the SingleInstructorFile tests are still trying to access `.modal-delete-button`, even though that was moved up to InstructorFiles. These tests likely just need to check that the right event is emitted.
  - Idea (not high priority): Instead of fixing the tests as mentioned, I would be ok with removing the trash can icon from SingleInstructorFile (and therefore these failing tests as well). I don't think there's a strong need to have two ways of deleting something.

**Medium Priority (not strictly required, but would be nice)**
- Idea: move the checkbox to the left of the filename (left/right padding in the SingleInstructorPanel outermost div `.single-instructor-file-component` should probably be reduced, may require some additional styling tweaks; the last modified timestamp doesn't need to be aligned with the filename).
- Idea: Always show the batch delete button, but make it disabled and gray) if no items are selected for deletion

caywagner committed:
Move SingleInstructorFile delete tests to InstructorFile

caywagner committed:
Revert auto formatting to clean diff logs

neiljohari committed:
Use arrays for reactivity, hide delete on collapse

Vue does not track changes on sets correctly. This commit switches to
arrays for reliably Vue interactions and updates checkbox binding
mechanism to use Vue sync.

The update to use Vue sync allows us to also restore checked files when
the sidebar is un-collapsed.

neiljohari committed:
Fix template tag issue and cleanup CSS

neiljohari committed:
Change toggle to be explicit by checkbox value

neiljohari committed:
Gray out and disable batch delete button if no items are selected

neiljohari committed:
Move checkbox to left of filename, improve style

neiljohari commented:
@james-perretta Hey James, we've tried making both the high and medium priority changes requested! 

![Screen Shot 2020-12-15 at 3 02 21 AM](https://user-images.githubusercontent.com/16354853/102187448-fe1dc800-3e81-11eb-8016-e17929c2e3a0.png)

![Screen Shot 2020-12-15 at 3 02 27 AM](https://user-images.githubusercontent.com/16354853/102187707-5654ca00-3e82-11eb-84e2-9b359fc32293.png)

Hopefully, this looks good. I also did some CSS tweak to improve the clickable area of the default checkbox on browsers that will choose to accept a scale transform on the size.

I tried it without the timestamp aligned and it didn't look very good, what do you think about this alignment where we fix the width of a checkbox and then align the timestamp to around that width?

james-perretta commented on src/components/project_admin/instructor_files/instructor_files.vue:
```diff
@@ -14,20 +14,35 @@
-          </span>
-          <span class="sidebar-header-text"
-                v-if="!d_collapsed || current_filename === null">Uploaded Files</span>
+          <div>
```
Remove this wrapping div, instead add `margin-left: auto` to `.batch-delete-files-button`

neiljohari replied:
ad641cf

james-perretta commented on src/components/project_admin/instructor_files/instructor_files.vue:
```diff
@@ -14,20 +14,35 @@
+            :key="instructor_file.pk"
+            :file="instructor_file"
+            @click="view_file(instructor_file)"
+            v-bind:selected="d_batch_to_be_deleted.some((f) => f.pk === instructor_file.pk)"
```
Change `v-bind` and `v-on` to the shorthand `:` and `@`

neiljohari replied:
2f8f255

james-perretta commented on src/components/project_admin/instructor_files/instructor_files.vue:
```diff
@@ -37,19 +52,66 @@
                    :progress="load_contents_progress"></view-file>
       </div>
     </div>
+    <div @click.stop>
```
What does this wrapping div with @click.stop do?

neiljohari replied:
Hm unknown, we got this line in the process of lifting your existing modal from here: https://github.com/eecs-autograder/ag-website-vue/blob/develop/src/components/project_admin/instructor_files/single_instructor_file.vue#L42

james-perretta replied:
Got it, we'll leave it in I can investigate later.

james-perretta commented on src/components/project_admin/instructor_files/instructor_files.vue:
```diff
@@ -76,6 +139,64 @@ export default class InstructorFiles extends OpenFilesMixin implements Instructo
+
+
+  // Returns whether we are in batch deletion mode
+  get batch_mode() {
```
Suggest renaming to `in_batch_mode`, `batch_mode_active`, or something similar

neiljohari replied:
e4406a8

james-perretta commented on src/components/project_admin/instructor_files/instructor_files.vue:
```diff
@@ -148,13 +269,14 @@ export default class InstructorFiles extends OpenFilesMixin implements Instructo
-@import '@/styles/collapsible_sidebar.scss';
+@import "@/styles/button_styles.scss";
+@import "@/styles/collapsible_sidebar.scss";
+@import "@/styles/forms.scss";
```
Does forms.scss get used in this file?

neiljohari replied:
e0ce707

neiljohari replied:
Oh I just realized, this _may_ be used for prettier buttons on the modal confirmation box. Add back?

neiljohari replied:
Added back in 0d342a3, can confirm its used by the modal

james-perretta commented on src/components/project_admin/instructor_files/instructor_files.vue:
```diff
@@ -181,34 +303,72 @@ $border-color: hsl(220, 40%, 94%);
-  padding-bottom: .5rem;
-  font-weight: bold;
-  font-size: 1.125rem;
+.batch-select-all-checkbox {
```
Does this exist? If not, remove

neiljohari replied:
e0ce707

james-perretta commented on src/components/project_admin/instructor_files/instructor_files.vue:
```diff
@@ -181,34 +303,72 @@ $border-color: hsl(220, 40%, 94%);
+      display: flex;
+      justify-content: space-between;
+      padding-top: .5rem;
+      padding-bottom: .5rem;
```
With `margin-left: auto;` added to `.batch-delete-files-button`, we can remove `justify-content`. `display: flex`, `padding-top`, and `padding-bottom` are redundant and can be removed (I think the padding is a holdover, but display: flex is new)

neiljohari replied:
e0ce707

james-perretta commented on src/components/project_admin/instructor_files/single_instructor_file.vue:
```diff
@@ -86,8 +62,8 @@ import ValidatedInput from '@/components/validated_input.vue';
-  make_error_handler_func
-} from '@/error_handling';
+  make_error_handler_func,
+} from "@/error_handling";
```
Revert quotes

neiljohari replied:
c601b99

james-perretta commented on src/components/project_admin/instructor_files/single_instructor_file.vue:
```diff
@@ -101,22 +77,30 @@ import { is_not_empty } from '@/validators';
   file!: InstructorFile;
 
+  @Prop({required: false, type: Boolean})
+  selected!: Boolean;
```
Suggest renaming to `selected_for_deletion` or `marked_for_deletion`

james-perretta commented on src/components/project_admin/instructor_files/single_instructor_file.vue:
```diff
@@ -101,22 +77,30 @@ import { is_not_empty } from '@/validators';
+  }
+
+  set selectedProxy(value: Boolean) {
+    this.$emit('update:selected', value);
```
Rename event to match name change to `selected`

neiljohari replied:
7467309

james-perretta commented on src/components/project_admin/instructor_files/single_instructor_file.vue:
```diff
@@ -21,53 +21,29 @@
-         @click.stop="download_file"></i>
-      <i class="far fa-trash-alt delete-file"
-         @click.stop="d_show_delete_instructor_file_modal = true"></i>
+      <input class="select-checkbox" @click.stop v-model="selectedProxy" type="checkbox" />
```
Replace v-model on `selectedProxy` with `:checked="..." and @change="$emit('...', $event)"`

neiljohari replied:
2f8f255

james-perretta commented on src/components/project_admin/instructor_files/single_instructor_file.vue:
```diff
@@ -101,22 +77,30 @@ import { is_not_empty } from '@/validators';
   d_download_progress: number | null = null;
   d_downloading = false;
 
+  get selectedProxy() : Boolean {
```
Remove `selectetProxy`

neiljohari replied:
2f8f255

james-perretta commented on src/components/project_admin/instructor_files/single_instructor_file.vue:
```diff
@@ -187,38 +158,50 @@ export function handle_rename_file_error(component: SingleInstructorFile, error:
-  .edit-file-name, .download-file, .delete-file {
-    margin: 0 .375rem;
-  }
+  .file-info-actions {
+    display: flex;
+    justify-content: space-between;
+    width: 90%;
```
If I remove the `file-info-actions` div and css class (and the `display`, `justify-content`, and `width` rules here), I don't see a difference in the rendered format. Am I missing something, or can this be removed?

neiljohari replied:
Looks like this for me with those attributes removed:

<img width="268" alt="Screen Shot 2020-12-18 at 6 54 48 PM" src="https://user-images.githubusercontent.com/16354853/102674043-92c64580-4162-11eb-9558-788bba9446d6.png">

The width 90% is to allow room for scrollbar. With this width, the icon is out of the way of the scroll bar but not enough to look visibly odd. Totally up to you whether you want to keep that, I just thought it worked better on a longer list of files.

james-perretta replied:
Got it. It looks fine either way in my browser (non-mac), so we'll keep it.

james-perretta commented on src/components/project_admin/instructor_files/single_instructor_file.vue:
```diff
@@ -236,28 +219,15 @@ export function handle_rename_file_error(component: SingleInstructorFile, error:
 
-.update-file-name-button, .update-file-name-cancel-button {
+.update-file-name-button,
+.update-file-name-cancel-button {
```
Revert formatting change

neiljohari replied:
c601b99

james-perretta commented on tests/test_project_admin/test_instructor_files/test_instructor_files.ts:
```diff
@@ -237,14 +237,96 @@ describe('InstructorFiles.vue', () => {
+
+        wrapper.findAll('.single-instructor-file-component input').at(0).trigger('click');
+        await wrapper.vm.$nextTick();
+
```
For this particular test, I think clicking on the checkbox just once is fine.
Suggest adding a test where you select multiple files for deletion, unselect a subset of those, then trigger the batch delete and make sure that the right files are getting deleted

neiljohari replied:
1f21b41

james-perretta reviewed (changes requested):
UI looks good! I left some suggestions for the source code. Also make sure that you run `npm run build` locally and fix the warnings that appear. Once those things are resolved, we can merge!

neiljohari committed:
Revert formatting changes

caywagner committed:
Move checkbox deselect test

neiljohari committed:
Replace wrapping div with margin-left

neiljohari committed:
Remove unused css

neiljohari committed:
Rename batch_mode to in_batch_mode

neiljohari committed:
Remove updated scope on custom events

neiljohari committed:
Remove selected proxy state and improve prop naming

neiljohari committed:
Add back forms css for modal buttons

neiljohari committed:
Adjust syntax for linter

james-perretta commented on tests/test_project_admin/test_instructor_files/test_instructor_files.ts:
```diff
@@ -280,6 +362,30 @@ describe('InstructorFiles.vue', () => {
+        wrapper.find('.delete-file').trigger('click');
+        await wrapper.vm.$nextTick();
+
+        wrapper.find('.modal-delete-button').trigger('click');
+        await wrapper.vm.$nextTick();
```
```suggestion
        await wrapper.find('.modal-delete-button').trigger('click');
        expect(await wait_until(wrapper, w => !w.vm.d_delete_pending)).toBe(true);
```

james-perretta commented:
Looks good! I left a suggestion that should fix that test failure. Go ahead and patch that in and we can merge!

neiljohari committed:
Update tests/test_project_admin/test_instructor_files/test_instructor_files.ts

Co-authored-by: James Perretta <jameslp@umich.edu>

james-perretta commented:
Awesome, can you rebase on develop?

neiljohari commented:
@james-perretta Rebased on develop and force pushed with the right history order

james-perretta commented:
@neiljohari Thanks again for this!

neiljohari commented:
@james-perretta Thanks for the opportunity! Was a great experience
